internal/controller/numaflowcontrollerrollout/numaflowcontrollerrollout_controller.go

```diff
@@ -501,20 +501,26 @@ func (r *NumaflowControllerRolloutReconciler) processNumaflowControllerStatus(
 			return err
 		}
 
-		healthyChildCond := existingNumaflowControllerStatus.GetCondition(apiv1.ConditionChildResourceHealthy)
-
-		if existingNumaflowControllerStatus.IsHealthy() &&
-			healthyChildCond != nil && existingNumaflowControllerDef.GetGeneration() <= healthyChildCond.ObservedGeneration &&
-			healthyChildCond.Status == metav1.ConditionTrue {
-
-			nfcRollout.Status.MarkChildResourcesHealthy(nfcRollout.Generation)
-		} else {
-			if healthyChildCond != nil {
-				nfcRollout.Status.MarkChildResourcesUnhealthy(healthyChildCond.Reason, healthyChildCond.Message, nfcRollout.Generation)
+		if existingNumaflowControllerDef.GetGeneration() > existingNumaflowControllerStatus.ObservedGeneration {
+			nfcRollout.Status.MarkChildResourcesUnhealthy("Progressing",
+				fmt.Sprintf("observedGeneration %d < generation %d", existingNumaflowControllerStatus.ObservedGeneration, existingNumaflowControllerDef.GetGeneration()),
+				nfcRollout.Generation)
+		} else if existingNumaflowControllerStatus.IsHealthy() {
+			// check the ChildResourcesHealthy Condition
+			healthyChildCond := existingNumaflowControllerStatus.GetCondition(apiv1.ConditionChildResourceHealthy)
+			if healthyChildCond != nil && healthyChildCond.Status == metav1.ConditionTrue {
+				nfcRollout.Status.MarkChildResourcesHealthy(nfcRollout.Generation)
 			} else {
-				nfcRollout.Status.MarkChildResourcesUnhealthy(apiv1.ProgressingReasonString, "Progressing", nfcRollout.Generation)
+				if healthyChildCond != nil {
+					nfcRollout.Status.MarkChildResourcesUnhealthy(healthyChildCond.Reason, healthyChildCond.Message, nfcRollout.Generation)
+				} else {
+					nfcRollout.Status.MarkChildResourcesUnhealthy("Unhealthy", "Unhealthy", nfcRollout.Generation)
+				}
 			}
+		} else {
+			nfcRollout.Status.MarkChildResourcesUnhealthy("Failed", "Failed", nfcRollout.Generation)
 		}
+
```

Comment:
Updated this function a little bit. I prefer to check the `NumaflowController.Status.ObservedGeneration` as opposed to the `ObservedGeneration` of the Condition to know whether NumaflowController has been reconciled